### PR TITLE
Remove notifications for deleted instances

### LINKED ISSUE #4845: Instance related notifications not cleared when instance deleted
### Current Behavior

If an instance crashes or enters safe mode, we send a notification to team owners.

If the instance is deleted, those notifications are not cleared; so the user still sees them and can click on them - which then 404s as the instance has been deleted.

### Expected Behavior

When an instance is deleted, clear any related notifications.

NB: due to the design of notifications, we can't lean on a delete cascade at the DB level; we need a post-delete hook in the Instance model to do the work.

As we will have notifications in this state already, we will need a migration to tidy up the table.

### Steps To Reproduce

- Crash an instance to generate a notification in the notification inbox
- Delete the instance
- Check the notification inbox - the notification should be removed

### Environment

- FlowFuse version:
- Node.js version:
- npm version:
- Platform/OS:
- Browser:

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### PR DESCRIPTION
**Pull request by hardillb**
fixes #4845

## Description

Adds afterDestroy hook to Projects table remove Notifications when Instances are deleted.

Also includes a migration to remove orphaned Notifications.

## Related Issue(s)

 #4845

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [x] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by hardillb** [comment]
We should probably do Devices as well, but want the approach verifying first

**Comment by knolleary** [comment]
Can we look at doing a bulk delete sql rather than lots of individual ones?

https://sequelize.org/docs/v7/querying/delete/#deleting-in-bulk

**Comment by hardillb** [comment]
I'll have another look, but we only have raw SQL access at this point (none of the models are loaded when the migrations run).

And the Notification table doesn't have a bare field with the project id that would could do a `WHERE "projectId" NOT IN (SELECT "id" from Projects)`

**Comment by knolleary** [comment]
Good point; the reference field doesn't lend itself to using with `in`.

File: forge/db/migrations/20241209-01-remove-dead-instance-notifications.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        for (const notification of notifications) {
+            const projectId = notification.reference.split(':')[1]
+            if (deletedProjects.includes(projectId)) {
+                // already found project to be gone, delete
+                deleteNotification(notification.id)
+            } else {
+                if (!foundProjects.includes(projectId)) {
+                    const projectQuery = `SELECT count(*) from "Projects" WHERE "id" = '${projectId}' LIMIT 1;`
+                    const [projectExits] = await context.sequelize.query(projectQuery)
+                    if (projectExits[0] === 0) {
```
Comment (knolleary) [inline forge/db/migrations/20241209-01-remove-dead-instance-notifications.js:30]: Running this against sqlite and postgres results in different shapes of results that don't work with the test that is then performed on them

**sqlite**
```
[ [ { 'count(*)': 1 } ], Statement {} ]
```

**postgres**
```
[ [ { count: '1' } ], Result { ... } ] ]
```


In neither case will the current code correctly match the actual result. Note the different field names and the values are different types (int/string).

```suggestion
                    const projectQuery = `SELECT count(*) as c from "Projects" WHERE "id" = '${projectId}' LIMIT 1;`
                    const [[projectCount]] = await context.sequelize.query(projectQuery)
                    const count = parseInt(projectCount.c)
                    if (count === 0) {
```

**Review by knolleary — CHANGES_REQUESTED** [review]
Issue with the count query across both support db flavours.

**Review by knolleary — APPROVED** [review]

**Pull request merged**